odoo/tools/_vendor/send_file.py

```diff
@@ -0,0 +1,235 @@
+"""
+Vendored copy of the werkzeug.utils.send_file function defined in
+werkzeug 2. It does not exist in werkzeug 0.16 which is the version we
+use.
+
+:copyright: 2007 Pallets
+:license: BSD-3-Clause
+"""
+
+import io
+import logging
+import mimetypes
+import os
+import typing as t
+import unicodedata
+from datetime import datetime
+from time import time
+from zlib import adler32
+
+from werkzeug.datastructures import Headers
+from werkzeug.exceptions import RequestedRangeNotSatisfiable
+from werkzeug.urls import url_quote
+from werkzeug.wrappers import Response
+from werkzeug.wsgi import wrap_file
+
+_logger = logging.getLogger(__name__)
+
+
+def send_file(
+    path_or_file: t.Union[os.PathLike, str, t.IO[bytes]],
+    environ: "WSGIEnvironment",
+    mimetype: t.Optional[str] = None,
+    as_attachment: bool = False,
+    download_name: t.Optional[str] = None,
+    conditional: bool = True,
+    etag: t.Union[bool, str] = True,
+    last_modified: t.Optional[t.Union[datetime, int, float]] = None,
+    max_age: t.Optional[
+        t.Union[int, t.Callable[[t.Optional[str]], t.Optional[int]]]
+    ] = None,
+    use_x_sendfile: bool = False,
+    response_class: t.Optional[t.Type["Response"]] = None,
+    _root_path: t.Optional[t.Union[os.PathLike, str]] = None,
+) -> "Response":
+    """Send the contents of a file to the client.
+    The first argument can be a file path or a file-like object. Paths
+    are preferred in most cases because Werkzeug can manage the file and
+    get extra information from the path. Passing a file-like object
+    requires that the file is opened in binary mode, and is mostly
+    useful when building a file in memory with :class:`io.BytesIO`.
+    Never pass file paths provided by a user. The path is assumed to be
+    trusted, so a user could craft a path to access a file you didn't
+    intend.
+    If the WSGI server sets a ``file_wrapper`` in ``environ``, it is
+    used, otherwise Werkzeug's built-in wrapper is used. Alternatively,
+    if the HTTP server supports ``X-Sendfile``, ``use_x_sendfile=True``
+    will tell the server to send the given path, which is much more
+    efficient than reading it in Python.
+    :param path_or_file: The path to the file to send, relative to the
```

Comment:
This needs paragraph breaks, and a lot of them. Outside of special constructs (e.g. info fields), rST only treats empty lines as block (and thus paragraph) separators, so here the entire docstring will be considered a single paragraph. Unless rST decides to parse the indents as quotes.

The original code [has proper paragraph breaks](https://github.com/pallets/werkzeug/blob/main/src/werkzeug/utils.py#L345-L363).